[quant] Custom module fix

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46229 [quant] Statically quantized LSTM
* #46426 [quant] Allow replacement of the custom modules in the "prepare" step
* **#46228 [quant] Custom module fix**
* #46232 [quant] Compute scale and zero point automatically in testing::_quantize
* #46235 [quant] Quantized flip dispatch

Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

Commit:
[quant] Custom module fix

[ghstack-poisoned]

Commit:
Update on "[quant] Custom module fix"




[ghstack-poisoned]

z-a-f commented:
Changes:
- Making the dictionaries `global` to make sure the scope of the variable is not limited
- Adding check if the modules are classes or instances to make sure we can check if an instance is custom as well

Commit:
Update on "[quant] Custom module fix"




[ghstack-poisoned]

Commit:
Update on "[quant] Custom module fix"


Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

[ghstack-poisoned]

Commit:
Update on "[quant] Custom module fix"


Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

[ghstack-poisoned]

Commit:
Update on "[quant] Custom module fix"


Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

[ghstack-poisoned]

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -9,15 +11,21 @@ def register_observed_custom_module_mapping(float_custom_module_class, observed_
     float custom module.
     """
+    global OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS
+    if float_custom_module_class is None:
+        return OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS
```
what's the context for adding this?  Seems unexpected, since this is changing the return type from a class to a mapping of classes.

z-a-f replied:
The point on the return type makes sense -- the motivation was for debugging purposes, sometimes I needed to see what was registered so far. There was no function for it, so I piggybacked on that function. I will revert it back

Commit:
Update on "[quant] Custom module fix"


Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

[ghstack-poisoned]

vkuzo commented on torch/quantization/quantize.py:
```diff
@@ -216,6 +217,14 @@ def prepare(model, inplace=False, allow_list=None,
+        if inplace:
+            warnings.warn("'inplace' does not apply to the custom modules")
+        model = get_observed_custom_module_class(model).from_float(model)
+        return prepare(
```
what's the context for using recursion here, and for doing it with inplace=True?  Can we avoid either/both?

z-a-f replied:
The process here is that the `get_observed_custom_module_class` replaces the model with a different implementation. After that we can prepare the newly instantiated model. The motivation for `inplace` is that because the newly created model is not the same as the original, there is no need to deepcopy it.

vkuzo reviewed (commented):
would it be possible to add some context to the PR summary (i.e. what's broken that this PR is fixing), and perhaps a test case to cover the new functionality?

z-a-f commented:
> would it be possible to add some context to the PR summary (i.e. what's broken that this PR is fixing), and perhaps a test case to cover the new functionality?

The first comment in this PR explains the motivation behind this PR. Do you need more context?

z-a-f commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -47,18 +54,37 @@ def get_quantized_custom_module_class(float_custom_module_class):
         ' does not have a corresponding quantized module class'
     return quantized_custom_module_class
 
+def is_custom_observed_module_class(module_class):
+    """ Check if a given module class is a custom observed module class
+    """
+    global OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS
+    if not inspect.isclass(module_class):
+        module_class = type(module_class)
+    return module_class in OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS
+
+def is_custom_quantized_module_class(module_class):
+    """ Check if a given module class is a custom quantized module class
+    """
+    global QUANTIZED_CUSTOM_MODULE_CLASS_MAPPINGS
+    if not inspect.isclass(module_class):
+        module_class = type(module_class)
+    return type(module_class) in QUANTIZED_CUSTOM_MODULE_CLASS_MAPPINGS
```
Just getters that will be needed later -- should I use a different names?

Commit:
Update on "[quant] Custom module fix"


Differential Revision: [D24270012](https://our.internmc.facebook.com/intern/diff/D24270012)

[ghstack-poisoned]
